File: application/spring-boot/src/main/java/codes/fdk/blueprint/api/application/spring/CategoryApiApplication.java
package codes.fdk.blueprint.api.application.spring;

import codes.fdk.blueprint.api.domain.service.CategoryService;
import codes.fdk.blueprint.api.domain.spi.CategoryRepository;
import codes.fdk.blueprint.api.infrastructure.persistence.r2dbc.R2dbcPersistenceModule;
import codes.fdk.blueprint.api.infrastructure.rest.webflux.RestApiConfig;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Import;

@SpringBootApplication
@Import({R2dbcPersistenceModule.class, RestApiConfig.class})
public class CategoryApiApplication {

    public static void main(String[] args) {
        SpringApplication.run(CategoryApiApplication.class, args);
    }

    @Bean
    public CategoryService categoryService(CategoryRepository categoryRepository) {
        return CategoryService.create(categoryRepository);
    }

}

File: application/vertx/src/main/kotlin/codes/fdk/blueprint/api/application/vertx/CategoryApiApplication.kt
package codes.fdk.blueprint.api.application.vertx

import codes.fdk.blueprint.api.domain.service.CategoryService
import codes.fdk.blueprint.api.infrastructure.vertx.eventbus.CategoryRepositoryEBProxy
import codes.fdk.blueprint.api.infrastructure.vertx.eventbus.CategoryServiceEBProxy
import codes.fdk.blueprint.api.infrastructure.vertx.eventbus.CategoryServiceEBProxyHandler
import codes.fdk.blueprint.api.infrastructure.vertx.persistence.postgres.PostgresPersistenceVerticle
import codes.fdk.blueprint.api.infrastructure.vertx.web.openapi.WebOpenapiVerticle
import io.vertx.config.ConfigRetriever
import io.vertx.core.CompositeFuture
import io.vertx.core.DeploymentOptions
import io.vertx.core.Vertx
import io.vertx.core.impl.logging.LoggerFactory
import io.vertx.core.json.JsonObject
import kotlin.system.exitProcess

private val log = LoggerFactory.getLogger("CategoryApiApplication")

//TODO health-check
fun main() {
    val vertx = Vertx.vertx()
    val categoryService = CategoryService.create(CategoryRepositoryEBProxy(vertx))

    vertx.eventBus()
        .localConsumer<JsonObject>(CategoryServiceEBProxy.ADDRESS)
        .handler(CategoryServiceEBProxyHandler(categoryService))

    ConfigRetriever.create(vertx)
        .config
        .compose {
            val deploymentOptions = DeploymentOptions().setConfig(it)
            CompositeFuture.all(
                vertx.deployVerticle(WebOpenapiVerticle(), deploymentOptions),
                vertx.deployVerticle(PostgresPersistenceVerticle(), deploymentOptions),
            )
        }
        .onSuccess { log.info("Deployment successful!") }
        .onFailure {
            log.error("Deployment failed!", it)
            exitProcess(-1)
        }
}
